lib/api/ldp/ldp-operation.js

```diff
@@ -0,0 +1,196 @@
+  'get': LdpGetOperation,
+  'put': LdpPutOperation,
+  'post': LdpPostOperation,
+  'patch': LdpPatchOperation,
+  'delete': LdpDeleteOperation,
+  'copy': LdpCopyOperation,
+  'options': LdpOptionsOperation
+}
+
+class LdpOperation {
+  /**
+   * @param options {object}
+   *
+   * @param options.target {URL} Parsed full request url
+   *
+   * @param options.headers {object}
+   *
+   * @param [options.authentication=null] {Authentication} Authn object,
+   *   contains WebID string, as well as any bearer credentials/tokens. Needed
+   *   for authenticated fetch (of remote group ACLs, of Copy resources, etc).
+   *   Null if request is not authenticated.
+   *
+   * @param [options.bodyStream] {Stream} Request body stream, for parsing when
+   *   needed
+   */
+  constructor (options) {
+    this.authentication = options.authentication
+    this.target = options.target
+    this.headers = options.headers
+    this.bodyStream = options.bodyStream
+  }
+
+  static async from ({req, host}) {
+    const { method, headers, session: { authentication } } = req
+    const target = LdpOperation.parseTargetUrl({req, host})
+    const bodyStream = req
+
+    const Operation = OPERATIONS[method]
+    if (!Operation) {
+      throw new HttpError(405, 'Method not supported')
+    }
+
+    return Operation.from({headers, target, bodyStream, authentication})
+  }
+
+  /**
+   * Taken from utils.getBaseUri()
+   *
+   * @param req {IncomingRequest}
+   * @param host {SolidHost}
+   *
+   * @returns {URL}
+   */
+  static parseTargetUrl ({req, host}) {
+    const protocol = host.serverUri.replace(/:.*/, '')
+
+    return url.format({
+      protocol: protocol || req.protocol,
+      host: req.get('host'),
+      pathname: url.resolve(req.baseUrl, req.path),
+      query: req.query
+    })
+  }
+
+  writeCommonHeaders ({res}) {
+    // set headers in common to all LDP responses
+  }
+}
+
+/**
+ * Checks to see if target exists
+ */
+class LdpHeadOperation extends LdpOperation {
+  writeHeaders ({res, result, permissions}) {
+    this.writeCommonHeaders({res})
+    // write HEAD-specific headers
+  }
+}
+
+/**
+ * Use cases to handle:
+ *  - "List Container" request, if target is an ldp container
+ *  - However, if Accept: header is HTML, check if index.html exists
+ *  - Need to also support similar use case if index.ttl exists
+ *  - "Glob pattern request" if glob magic chars are detected (ie `*`)
+ *  - Otherwise, plain Get request
+ */
+class LdpGetOperation extends LdpOperation {
+  static from ({headers, target, bodyStream, authentication, store}) {
+    const targetUrl = url.format(target)
+
+    if (glob.hasMagic(targetUrl)) {
+      // this is a Ldp Glob operation, return that
+    }
+
+    // check to see if target is a container (based on Link: type rel)
+    // ...
+    // if it is a container, check to see if index.html exists
+    const indexFile = url.resolve(targetUrl, '/index.html')
+    if (store.exists(indexFile)) {
```

Comment:
@RubenVerborgh - the GET request dispatching logic here illustrates why I was thinking it makes sense to include the `store` (backend) param to `Operation.from()`.

Reply:
Mmm, I think the `from` is doing at lot here. [Constructors shouldn't do actual work](http://misko.hevery.com/code-reviewers-guide/flaw-constructor-does-real-work/) (and although a static method, `from` is acting as a constructor here).

I think the constructor should just create/parse the operation, and then we have a separate `operation.perform(store)` method.

The question probably is how much store-agnostic work an Operation can do.

Reply:
Ok, I also believe that constructors shouldn't do actual work. But as far as I know, factory methods that set up and initialize the various things to pass to a constructor, are a valid strategy. So the `from()` method is just a factory method.

Now, as far as whether a store is involved in *defining* the operation - so my point here is that without a store, it's impossible to determine the operation. (Specifically, to decide whether this is a List Container op or a Get Op).

Reply:
> But as far as I know, factory methods that set up and initialize the various things to pass to a constructor, are a valid strategy. So the from() method is just a factory method.

It's never a black-and-white situation, but I perceive the static `from` method more as a de-facto constructor rather than a factory method:

1. I is a static method, and constructors are (kind of) static methods too. The idea of the factory pattern is precisely to not depend on static methods.
2. The idea of a factory is that users of the factory don't need to know how the factory creates its objects. Some other component has instantiated a factory with several constructor arguments, so when we call a factory method, we don't need to pass in lots of things over and over again (like protocol and host etc.). I.e., the nice thing of a factory is that it is an instantiated object, so we don't need to pass all the state on every method call.
3. There is no practical difference (from a caller's perspective) if we move the code currently in  `from` to the constructor.

> Now, as far as whether a store is involved in defining the operation - so my point here is that without a store, it's impossible to determine the operation. 

Depending on the definition of “operation”, you do have a valid point here. Perhaps should be updated in the architecture document.